Multi variate

agneborn98 opened the pull request:
This PR adds support for creating a record builder from any schema retrieved through `GetSchema`.

Currently, the plugin expects any string found in the schema to be a tag from the telegraf metric, as I couldn't think of a situation where a string in the schema would be a field in the metric and not a tag. If you don't agree, please advise.

As you can see in the following image, the output plugin is currently able to send metrics at a rate of about one metric every 10-15 microseconds:
![image](https://user-images.githubusercontent.com/43062786/187199670-0b25d453-e109-4366-ab2c-f88ffdc25e7b.png)

agneborn98 committed:
First iteration Multi-variate

agneborn98 committed:
Second iteration multivariate and dynamic builders

agneborn98 committed:
Removed prints

agneborn98 committed:
Update flight.go

agneborn98 committed:
Full support for any schema

This commit adds full support for any schema retrieved through get_schema() and includes support for multivariate schemas

CGodiksen commented on plugins/output/flight/flight.go:
```diff
@@ -47,27 +48,86 @@ func (f *Flight) Write(metrics []telegraf.Metric) error {
 	builder := array.NewRecordBuilder(memory.DefaultAllocator, &f.schema)
```
The "Write()" function could use some more comments in general.

agneborn98 replied:
Is this okay or do you think I should do more comments?

CGodiksen commented on plugins/output/flight/flight.go:
```diff
@@ -47,27 +48,86 @@ func (f *Flight) Write(metrics []telegraf.Metric) error {
-
-		valueBuilder.AppendValues([]float32{float32(fieldlst[0].Value.(float64))}, nil)
+		for i, f := range schemaFields {
+			switch f.Type.ID() {
```
Are all possible types represented here?

agneborn98 replied:
No, I guess I wanted some input from other people on what kinds of types were deemed to be necessary. The [list](https://pkg.go.dev/github.com/apache/arrow/go/v9@v9.0.0/arrow#Type) is REALLY extensive, but I think I included those I thought would appear in a schema. Any thoughts?

CGodiksen replied:
I am not sure what would be best. What do you think @skejserjensen and @chrthomsen?

skejserjensen replied:
I think we should try to support all of the different types if at all possible. However, to make the code readable and maintainable may consider using `go generate` to automatically generate the code for each type at compile time? For example, how `go generate` can be used to run [Yacc](https://go.dev/blog/generate) and Eli Bendersky's website for a demonstration of some of the [more advanced features](https://eli.thegreenplace.net/2021/a-comprehensive-guide-to-go-generate/). If code-generation cannot be used the switches body should probably be moved to a separate function to make `Write()` simpler to read.

agneborn98 replied:
Can we push the code generation to another Pull Request? I agree that it is definitely the more elegant solution, but I think it is a bit much to change the entire code base in this from what was initially requested. On top of that, the performance of the current Write function is clearly enough to satisfy the KPI in the MORE project. What do you think @skejserjensen ?

skejserjensen replied:
Fine for me.

CGodiksen commented on plugins/output/flight/flight.go:
```diff
@@ -47,27 +48,86 @@ func (f *Flight) Write(metrics []telegraf.Metric) error {
-		fieldlst := m.FieldList()
-
-		valueBuilder.AppendValues([]float32{float32(fieldlst[0].Value.(float64))}, nil)
+		for i, f := range schemaFields {
```
Maybe use "field" instead of "f".

agneborn98 replied:
Fixed.

CGodiksen commented on plugins/output/flight/flight.go:
```diff
@@ -47,27 +48,86 @@ func (f *Flight) Write(metrics []telegraf.Metric) error {
 
 	builder.Reserve(len(metrics))
 
 	for _, m := range metrics {
```
Maybe use "metric" instead of "m".

agneborn98 replied:
Fixed.

CGodiksen commented on plugins/output/flight/flight.go:
```diff
@@ -47,27 +48,86 @@ func (f *Flight) Write(metrics []telegraf.Metric) error {
+			case arrow.TIMESTAMP:
+				builder.Field(i).(*array.TimestampBuilder).AppendValues([]arrow.Timestamp{arrow.Timestamp(timeInt)}, nil)
+			case arrow.STRING:
+				tag, wasSet := m.GetTag(f.Name)
```
Maybe add some comments to explain why "m.GetTag()" and "m.GetField" is necessary.

agneborn98 replied:
I think this has been addressed in the comment now.

CGodiksen commented on plugins/output/flight/flight.go:
```diff
@@ -47,27 +48,86 @@ func (f *Flight) Write(metrics []telegraf.Metric) error {
+				if !wasSet {
+					fmt.Printf("field %d : %s not set", i, f.Name)
+				}
+				switch v := field.(type) {
```
Why it is necessary to go through the types again inside each switch case? Maybe add a comment explaining this.

agneborn98 replied:
I think this has been addressed in the comment now.

CGodiksen commented on plugins/output/flight/flight.go:
```diff
@@ -47,27 +48,86 @@ func (f *Flight) Write(metrics []telegraf.Metric) error {
+				if !wasSet {
+					fmt.Printf("field %d : %s not set", i, f.Name)
+				}
+				switch v := field.(type) {
```
There is a lot of duplicated code in these internal switch statements. Go should have support for multiple case values. For example, here "int64, float32, float64" could be the same case.

agneborn98 replied:
You are right, Go supports multiple case values. But the switch here is a type switch, that in some ways does not support multiple case values. When a type assertion is made, this assertion is only used on the first case in the multiple cases entered, but the rest of the cases are reverted back to interfaces. As I cannot insert an interface into the builder, I don't see a workaround for this. For more information, see [this](https://stackoverflow.com/questions/40575033/golang-multiple-case-in-type-switch).

![image](https://user-images.githubusercontent.com/43062786/187420724-9dd7cd28-a95b-46ba-a57f-e9dc0d2c1b9e.png)

CGodiksen replied:
Could you use an if statement instead? This would make it possible to have multiple boolean expressions in a single check to group the duplicate case handlers.

skejserjensen replied:
As I mentioned in this [comment](https://github.com/ModelarData/Telegraf-Output-Apache-Arrow-Flight/pull/2#discussion_r958565365), maybe `go generate` can be used to reduce the amount of duplicate code.

agneborn98 replied:
That is true, I would be able to use an if statement here. The problem is that I have to do a type assertion to extract the value from the interface, and I can't do three separate type assertions within one if else statement. This is exactly why type switches [were](https://go.dev/tour/methods/16) [made](https://stackoverflow.com/a/6373523).

![image](https://user-images.githubusercontent.com/43062786/187499773-bfa3072f-7cc5-4a8b-891a-d36bb34f8017.png)

As you can see in this image, I have no idea how I would approach the type in the question marks.

CGodiksen commented on plugins/output/flight/flight.go:
```diff
@@ -47,27 +48,86 @@ func (f *Flight) Write(metrics []telegraf.Metric) error {
+					fmt.Printf("tag %d : %s not set", i, f.Name)
+				}
+				builder.Field(i).(*array.StringBuilder).AppendValues([]string{tag}, nil)
+			case arrow.INT32:
```
it seems that the almost duplicated code in the field cases could be removed by using a helper function.

agneborn98 replied:
I think this is fixed now, was this what you had in mind?

CGodiksen replied:
I was more thinking if it was possible to put the entire internal switch statement into a function. The function would then take the builder and type as arguments.

skejserjensen replied:
I agree with @CGodiksen that the switch should probably be moved to a separate function if it cannot be generated.

agneborn98 replied:
I have moved the entire loop with switches to another function with the minimal amount of parameters I could think of. Is this okay?

CGodiksen replied:
It does not really solve the problem of the internal switch cases being large duplicated code. The intention was that all the field handlers could use a single function instead of duplicating this code:

```
metricField := getField(metric, schemaField, i)
switch value := metricField.(type) {
case int32:
	builder.Field(i).(*array.Int32Builder).AppendValues([]int32{value}, nil)
case int64:
	builder.Field(i).(*array.Int32Builder).AppendValues([]int32{int32(value)}, nil)
case float32:
	builder.Field(i).(*array.Int32Builder).AppendValues([]int32{int32(value)}, nil)
case float64:
	builder.Field(i).(*array.Int32Builder).AppendValues([]int32{int32(value)}, nil)
}
```

If the choice to use code generation is taken this might not be relevant.

agneborn98 replied:
I see.. Yeah, I can give you a whole writeup of why that type assertion (`(*array.Int32Builder)`) is very comlicated to implement into a seperate function without dynamic code generation. 
Instead, I think this should just be left alone for now, and postponed until a final design for this entire builder process has been made.

CGodiksen replied:
Makes sense.

agneborn98 committed:
Update based on comments

CGodiksen commented on plugins/output/flight/flight.go:
```diff
@@ -44,38 +45,112 @@ type Flight struct {
 	return f.writer.Write(rec)
 }
 
+// getTag returns the value of the metric tag with the name equal to the name of the schema field.
```
Maybe just "Return the value of the metric tag with the name equal to the name of the schema field."

agneborn98 replied:
Fixed.

CGodiksen commented on plugins/output/flight/flight.go:
```diff
@@ -44,38 +45,112 @@ type Flight struct {
+	return metricTag
+}
+
+// getField returns the value of the metric field with the name equal to the name of the schema field.
```
Maybe just "Return the value of the metric field with the name equal to the name of the schema field."

agneborn98 replied:
Fixed.

agneborn98 committed:
Update flight.go

Switches in response to the comments, and switched some functions around for clarity in the order that they are executed.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -131,11 +99,124 @@ func (f *Flight) Connect() error {
+
+	// Iterate through the metrics and add them to the RecordBuilder.
+	for _, metric := range metrics {
+
```
Very minor, but maybe remove this empty line to be consistent with the loop in `addMetricToRecordBuilder`.

agneborn98 replied:
I think this is fixed now, was this what you meant?

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -131,11 +99,124 @@ func (f *Flight) Connect() error {
+	for i, schemaField := range schemaFields {
+		switch schemaField.Type.ID() {
+		case arrow.TIMESTAMP:
+			builder.Field(i).(*array.TimestampBuilder).AppendValues([]arrow.Timestamp{arrow.Timestamp(timeInt)}, nil)
```
Can [`.AppendValues()`](https://pkg.go.dev/github.com/apache/arrow/go/arrow/array#TimestampBuilder.AppendValues) be replaced with[ `.Append()`](https://pkg.go.dev/github.com/apache/arrow/go/arrow/array#TimestampBuilder.Append) everywhere to not create a slice for each element that is added to the builders?

agneborn98 replied:
Fixed.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -131,11 +99,124 @@ func (f *Flight) Connect() error {
+// as floats. If the schema expects an integer in the field, Apache Arrow will panic.
+func addMetricToRecordBuilder(schemaFields []arrow.Field, builder *array.RecordBuilder, metric telegraf.Metric) {
+
+	timeInt := metric.Time().UnixMilli()
```
Since `timeInt` is only used on [Line 150](https://github.com/ModelarData/Telegraf-Output-Apache-Arrow-Flight/blob/multi-variate/plugins/output/flight/flight.go#L150), maybe add it between Line 149 and Line 150 like the calls to `getTag()` and `getField()` so it is as close to where it is used as possible.

agneborn98 replied:
Fixed.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -131,11 +99,124 @@ func (f *Flight) Connect() error {
+func getTag(metric telegraf.Metric, schemaField arrow.Field, i int) string {
+	metricTag, wasSet := metric.GetTag(schemaField.Name)
+	if !wasSet {
+		panic(fmt.Sprintf("tag %d : %s not set", i, schemaField.Name))
```
Maybe add a comment to document why `panic()` is used instead of returning an error in `Write()`. I assume it is because there is no reason for Telegraf to try and retransmit the data if a required value is missing.

agneborn98 replied:
Fixed.

skejserjensen replied:
Looks like "a retransmit" is duplicated and that "recuired" should be "required". Maybe write "This is to prevent Telegraf from attempting to retransmit `metric` when the required value is missing".

agneborn98 replied:
Fixed.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -131,11 +99,124 @@ func (f *Flight) Connect() error {
+func getField(metric telegraf.Metric, schemaField arrow.Field, i int) interface{} {
+	metricField, wasSet := metric.GetField(schemaField.Name)
+	if !wasSet {
+		panic(fmt.Sprintf("field %d : %s not set", i, schemaField.Name))
```
Maybe add a comment to document why `panic()` is used instead of returning an error in `Write()`. I assume it is because there is no reason for Telegraf to try and retransmit the data if a required value is missing.

agneborn98 replied:
Fixed.

skejserjensen replied:
Looks like "recuired" should be "required". If the above comment is changed to "This is to prevent Telegraf from attempting to retransmit `metric` when the required value is missing" then this comment should probably be changed as well.

agneborn98 replied:
Fixed.

agneborn98 committed:
Fixes to @skejserjensen comments

agneborn98 committed:
Update flight.go

agneborn98 committed:
Update flight.go
